Add UUID best practices (#7640)
pingcap/docs: best-practices/uuid.md

---
title: UUID Best Practices
summary: Learn best practice and strategy for using UUIDs with TiDB.
---

# UUID Best Practices

## Overview of UUIDs

When used as a primary key, instead of an `AUTO_INCREMENT` integer value, a universally unique identifier (UUID) delivers the following benefits:

> - UUIDs are designed to be globally unique, and can be generated on multiple systems without risking conflicts. In some cases, this means that the number of network trips to TiDB can be reduced, leading to improved performance.

Reviewer: ```suggestion
- UUIDs can be generated on multiple systems without risking conflicts. In some cases, this means that the number of network trips to TiDB can be reduced, leading to improved performance.
```

- UUIDs are supported by most programming languages and database systems.

> - When used as a part of a URL, a UUID is not vulnerable to enumeration attacks. In comparison, with an `auto_increment` number, it is possible to guess the other invoice IDs or user IDs.

Reviewer: ```suggestion
- When used as a part of a URL, a UUID is not vulnerable to enumeration attacks. In comparison, with an `auto_increment` number, it is possible to guess the invoice IDs or user IDs.
```

## Best practices

### Store as binary

> The textual UUID format looks like this: `ab06f63e-8fe7-11ec-a514-5405db7aad56`, this is a string of 36 characters. By using `UUID_TO_BIN()` the textual format can be converted into a binary format of 16 bytes. This allows you to store this in `BINARY(16)` column. When retrieving you can use the `BIN_TO_UUID()` function to get back to the textual format.

Author: ```suggestion
The textual UUID format looks like this: `ab06f63e-8fe7-11ec-a514-5405db7aad56`, which is a string of 36 characters. By using `UUID_TO_BIN()`, the textual format can be converted into a binary format of 16 bytes. This allows you to store the text in a `BINARY(16)` column. When retrieving the UUID, you can use the `BIN_TO_UUID()` function to get back to the textual format.
```

### UUID format binary order and a clustered PK

The `UUID_TO_BIN()` function can be used with one argument, the UUID or with two arguments where the second argument is a `swap_flag`. It is recommended to not set the `swap_flag` with TiDB to avoid [hotspots](/best-practices/high-concurrency-best-practices.md).

To avoid hotspots it is recommended to explicitly set [the `CLUSTERED` option](/clustered-indexes.md) for UUID based primary keys to avoid hotspots.

To demonstrate the effect of the `swap_flag`, here are two tables with an identical structure. The difference is that the data inserted into `uuid_demo_1` uses `UUID_TO_BIN(?, 0)` and `uuid_demo_2` uses `UUID_TO_BIN(?, 1)`.

> In the screenshot of the [Key Visualizer](/dashboard/dashboard-key-visualizer.md) below you can see that writes are concentrated in a single region of the `uuid_demo_2` table that has the order of the fields swapped in the binary format.

Reviewer: ```suggestion
In the screenshot of the [Key Visualizer](/dashboard/dashboard-key-visualizer.md) below,
 you can see that writes are concentrated in a single region of the `uuid_demo_2` table that has the order of the fields swapped in the binary format.
```

![Key Visualizer](/media/best-practices/uuid_keyviz.png)

```sql
CREATE TABLE `uuid_demo_1` (
  `uuid` varbinary(16) NOT NULL,
  `c1` varchar(255) NOT NULL,
  PRIMARY KEY (`uuid`) CLUSTERED
)
```

```sql
CREATE TABLE `uuid_demo_2` (
  `uuid` varbinary(16) NOT NULL,
  `c1` varchar(255) NOT NULL,
  PRIMARY KEY (`uuid`) CLUSTERED
)
```

> ## MySQL Compatibility

Reviewer: ```suggestion
## MySQL compatibility
```

> UUID's can be used in MySQL as well. The `BIN_TO_UUID()` and `UUID_TO_BIN()` functions were introduced in MySQL 8.0. The `UUID()` function is available in older MySQL versions as well.

Reviewer: ```suggestion
UUIDs can be used in MySQL as well. The `BIN_TO_UUID()` and `UUID_TO_BIN()` functions were introduced in MySQL 8.0. The `UUID()` function is available in earlier MySQL versions as well.
```

---
title: UUID Best Practices
summary: Learn best practice and strategy for using UUID's with TiDB.
---

# UUID Best Practices

## The usecase for UUID

> With TiDB and MySQL UUID's are often used as primary key instead of a integer primary key.
> 
> The benefits of UUID's are:
> - UUID's are designed to be globally unique they can be generated on multiple systems without risking conflicts.
> - UUID's are supported by most programming languages and database systems

Reviewer 2: I'm not sure I'd say _often_ because it might sound contradictory to other advice. But I'd be okay with saying might be used, or something less common.

```suggestion

There are several benefits to using a `UUID` as a primary key, instead of an `AUTO_INCREMENT` integer value:

- They are designed to be globally unique they can be generated on multiple systems without risking conflicts. In some cases, this means that the number of network trips to TiDB can be reduced, leading to improved performance.
- They are supported by most programming languages and database systems.
- When used as part of URLs, a UUID is not vulnerable to enumeration attacks. i.e. with an `auto_increment` number, it is possible to guess the other invoice ids or user ids.
```

Author: Thanks!

> There are several benefits to using a `UUID` as a primary key, instead of an `AUTO_INCREMENT` integer value:
> 
> - They are designed to be globally unique they can be generated on multiple systems without risking conflicts. In some cases, this means that the number of network trips to TiDB can be reduced, leading to improved performance.
> - They are supported by most programming languages and database systems.
> - When used as part of URLs, a UUID is not vulnerable to enumeration attacks. i.e. with an `auto_increment` number, it is possible to guess the other invoice ids or user ids.

Reviewer 2: I think it is slightly easier to read as:
```suggestion
There are several benefits to using a `UUID` as a primary key, instead of an `AUTO_INCREMENT` integer value:

- UUIDs are designed to be globally unique, and can be generated on multiple systems without risking conflicts. In some cases, this means that the number of network trips to TiDB can be reduced, leading to improved performance.
- UUIDs are supported by most programming languages and database systems.
- When used as part of URLs, a UUID is not vulnerable to enumeration attacks. i.e. with an `auto_increment` number, it is possible to guess the other invoice ids or user ids.
```

Reviewer 2: Would you also say that an advantage of UUIDs is that in some cases it is less likely to hotspot than an `AUTO_INCREMENT` value? It really depends of course...

Author: Not sure about hotspots. That indeed depends on many things. Let's leave it out for now.

## Using UUID's with TiDB

> In TiDB you can generate UUID's with [`UUID()`](/functions-and-operators/miscellaneous-functions.md) and use `UUID_TO_BIN()` and `BIN_TO_UUID()` to convert from a text format to a binary format and back.

Author: done
